contrib/ruby/lib/trilogy.rb

```diff
@@ -1,7 +1,79 @@
-require "trilogy/cext"
 require "trilogy/version"
 
 class Trilogy
+  # Trilogy::Error is the base error type. All errors raised by Trilogy
+  # should be descendants of Trilogy::Error
+  module Error
+  end
+
+  # Trilogy::ConnectionError is the base error type for all potentially transient
+  # network errors.
+  module ConnectionError
+    include Error
+  end
+
+  class BaseError < StandardError
+    include Error
+  end
+
+  # Trilogy::ClientError is the base error type for invalid queries or parameters
+  # that shouldn't be retried.
+  class ClientError < BaseError
+    include Error
+  end
+
+  class QueryError < ClientError
+  end
+
+  class CastError < ClientError
+  end
+
+
+  class TimeoutError < Errno::ETIMEDOUT
+    include ConnectionError
+  end
+
+  class ProtocolError < BaseError
+    ERROR_CODES = {
+      1205 => TimeoutError, # ER_LOCK_WAIT_TIMEOUT
+      1044 => ConnectionError, # ER_DBACCESS_DENIED_ERROR
```

Comment:
I just realized. We can't raise `ConnectionError` as it's a module. We'll need to create another error that include that module.

We should also probably try to cover that path.